### ISSUE TITLE
Proposal: new --restart=daemon to cover only daemon start/stop cases

### ISSUE THREAD

**Issue by gdm85** [docker/docker#8821]
I am looking for a way to restore containers with their HostConfig (network etc.) in case the Docker daemon was rebooted.

Current feature
==================
Specifically, I cannot use ``--restart=always`` because - although I would like to keep their network configuration across a daemon restart - during the lifecycle of a healthy docker daemon service I have following possible cases:
- some containers can and will fail (unexpectedly)
- some containers need to be inspected by a (human) operator in case of failure
- if the daemon keeps trying to restart a container, you do not see it as "Exited" in the ``docker ps`` output, which is misleading, yet nobody can run anything on them using ``docker exec`` because they are in this continuous loop until it's given up

In general I find #7226 to not be (in its current form) in favor of strict devops practices, but rather allowing a loose approach like: if something fails, try again (instead of going to fix the reason of failure in first place). This can lead to catastrophic failures later on if nobody is following the number of restarts, that might become more and more frequent if problem does not get addressed (imagine increasing failures in an external/internal service etc).

This collides with many development and production use-cases I have been trying to model with Docker.

Missing option
============

To recap: I am looking for a feature where containers are restarted **only** in case of a daemon startup operation, and never in any other case. I tried to temptatively name such new option ``--restart=daemon`` to signal that a restart would be decided only based on daemon status.

P.S. if what I aim to do (restoring network configuration on a "start") is possible without using a ``--restart`` policy, please explain me instead :)

**Comment by SvenDowideit** [comment]
also see Container should not restart as a result of outside container errors: 
https://groups.google.com/forum/#!topic/docker-user/Ba2t9Qfzxw0

**Comment by gdm85** [comment]
@SvenDowideit yes the problem of Matt Hughes is more or less the same I'd like to exclude from happening in production environments

**Comment by SvenDowideit** [comment]
mmm, @crosbymichael you were the one that made the new restart-policy - is there a tweak we can have?

**Comment by gdm85** [comment]
With recent additions to [docker-fw](https://github.com/gdm85/docker-fw), I have at least solved my use-case.

Please also note that I have implemented topological sorting of the container/links [DAG](https://en.wikipedia.org/wiki/Directed_acyclic_graph), instead of the current approach with ``--restart=always`` that basically goes like: "try to restart them until all of them are started" ([relevant code](https://github.com/gdm85/docker-fw/blob/master/graph.go)).

Issue is **not solved** but I provide an (external) workaround.

**Comment by gdm85** [comment]
Am I correct to think that part of this feature was "predated" by #11008 and its merged PR #15348? Perhaps scope needs to be redefined...

**Comment by icecrime** [comment]
@gdm85 More importantly than the PR you linked, Docker 1.12 introduced the ability to _not_ kill containers when the daemon gets restarted. This is an opt-in feature called `--live-restore` which is documented [here](https://docs.docker.com/engine/admin/live-restore/).

Please let us know if something is still missing!

**Comment by gdm85** [comment]
@icecrime thanks for mentioning that, I heard about it but didn't know the name.

I am closing this because I would need to reformulate my use-case with current Docker features to see if there are blind spots left that need work on.

Thanks all

**Comment by francislavoie** [comment]
I think this issue still describes what I want, but doesn't yet exist.

I have a container that does a task at start-up which contacts a remote server which has rate limiting. If this fails too many times, the rate limit gets hit and I get stuck. I use `unless-stopped` which works great for me if the container is correctly configured, but if there's a problem, it'll very quickly hit the rate limits because it's constantly restarting.

What I want is a restart policy that only restarts the container on daemon/system restart, and NOT on container errors (or maybe a configurable max-restarts attempts on error, defaulting to 0, in the best-case).
